CodeEvolution/configs.txt
{'size': 150, 'population': Population(ID=2, proportion=1), 'mutant': Population(ID=8, proportion=0), 'socialNormID': 2, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 5000, 'updateProbability': 0.1, 'delta': 1, 'mutationProbability': 0.1, 'density': 1, 'degree': None, 'attachment': None, 'smallWorld': None}

{'size': 150, 'population': Population(ID=3, proportion=1), 'mutant': Population(ID=8, proportion=0), 'socialNormID': 3, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 5000, 'updateProbability': 0.1, 'delta': 1, 'mutationProbability': 0.1, 'density': 1, 'degree': None, 'attachment': None, 'smallWorld': None}

{'size': 150, 'population': Population(ID=4, proportion=1), 'mutant': Population(ID=8, proportion=0), 'socialNormID': 4, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 5000, 'updateProbability': 0.1, 'delta': 1, 'mutationProbability': 0.1, 'density': 1, 'degree': None, 'attachment': None, 'smallWorld': None}

{'size': 150, 'population': Population(ID=5, proportion=1), 'mutant': Population(ID=8, proportion=0), 'socialNormID': 5, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 5000, 'updateProbability': 0.1, 'delta': 1, 'mutationProbability': 0.1, 'density': 1, 'degree': None, 'attachment': None, 'smallWorld': None}



Experiments/Exp12/configs.txt
{'size': 300, 'population': Population(ID=0, proportion=1.0), 'mutant': Population(ID=8, proportion=0.0), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.95), 'mutant': Population(ID=8, proportion=0.05), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.9), 'mutant': Population(ID=8, proportion=0.1), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.85), 'mutant': Population(ID=8, proportion=0.15), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.8), 'mutant': Population(ID=8, proportion=0.2), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.75), 'mutant': Population(ID=8, proportion=0.25), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.7), 'mutant': Population(ID=8, proportion=0.3), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.65), 'mutant': Population(ID=8, proportion=0.35), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.6), 'mutant': Population(ID=8, proportion=0.4), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.55), 'mutant': Population(ID=8, proportion=0.45), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.5), 'mutant': Population(ID=8, proportion=0.5), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.45), 'mutant': Population(ID=8, proportion=0.55), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.4), 'mutant': Population(ID=8, proportion=0.6), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.35), 'mutant': Population(ID=8, proportion=0.65), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.3), 'mutant': Population(ID=8, proportion=0.7), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.25), 'mutant': Population(ID=8, proportion=0.75), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.2), 'mutant': Population(ID=8, proportion=0.8), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.15), 'mutant': Population(ID=8, proportion=0.85), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.1), 'mutant': Population(ID=8, proportion=0.9), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}

{'size': 300, 'population': Population(ID=0, proportion=0.05), 'mutant': Population(ID=8, proportion=0.95), 'socialNormID': 0, 'socialDilemma': {'name': 'PrisonersDilemma', 'benefit': 2, 'cost': 1}, 'omega': 0.99, 'maxPeriods': 3000, 'updateProbability': 0.1, 'delta': 0.5, 'mutationProbability': 0, 'density': 0.03802521649770801, 'degree': None, 'attachment': 2, 'smallWorld': None}



CodeEvolution/config.py
from CodeEvolution.socialdilemna import PrisonersDilemma

from collections import namedtuple
import numpy as np
import pickle

Dilemma = namedtuple('Dilemma', ['type', 'benefit', 'cost'])
State = namedtuple('State', ['mainID', 'proportion', 'mutantID'])
Population = namedtuple('Population', ['ID', 'proportion'])


class Config:
    """Configuration file describing a single parametrization of a simulation """

    def __init__(self,
                 initialState: State = State(mainID=0, proportion=0.9, mutantID=8),
                 size: int = 500,
                 densities: float = 1,
                 socialDilemma: Dilemma = Dilemma(type='PD', benefit=2, cost=1),
                 omegas: float = 0.99,
                 maxPeriods: int = 2000,
                 updateProbability: float = 0.1,
                 delta: float = 0.5,
                 mutationProbability: float = 0.1,
                 sparseDensity: bool = False,
                 degree: int = None) -> None:

        # TODO Validation checks (check validator class, should be able to just copy over)
        self.size = size
        self.population = Population(ID=initialState.mainID, proportion=initialState.proportion)
        self.mutant = Population(ID=initialState.mutantID, proportion=round(1 - initialState.proportion, 4))
        self.socialNormID = initialState.mainID  # Use the same social norm ID as the main strategy
        self.socialDilemma = PrisonersDilemma(socialDilemma.benefit, socialDilemma.cost)
        self.omega = omegas
        self.maxPeriods = maxPeriods
        self.updateProbability = updateProbability
        self.delta = delta
        self.mutationProbability = mutationProbability
        if sparseDensity:
            self.density = 2 * np.log(size) / size
        else:
            self.density = densities
        self.degree = degree

        if self.population.proportion + self.mutant.proportion != 1:
            total = self.population.proportion + self.mutant.proportion
            raise Exception(f"Population proportions sum to {total}, must sum to 1.")

    def __str__(self):
        s = ""
        for thing in self.__dict__.items():
            s += str(thing) + "\n"
        return s

    def __repr__(self):
        return str(self.__dict__)


if __name__ == "__main__":

    C = Config(initialState=State(mainID=7, proportion=1, mutantID=9))
    print(C)
